src/library/scala/collection/TraversableLike.scala

```diff
@@ -451,17 +451,23 @@ trait TraversableLike[+A, +Repr] extends Any
   }
 
   def groupBy[K](f: A => K): immutable.Map[K, Repr] = {
-    val m = mutable.Map.empty[K, Builder[A, Repr]]
-    for (elem <- this) {
-      val key = f(elem)
-      val bldr = m.getOrElseUpdate(key, newBuilder)
+    object m extends mutable.HashMap[K, Builder[A, Repr]] {
+      override def entriesIterator: Iterator[mutable.DefaultEntry[K, Builder[A, Repr]]] =
+        super.entriesIterator
+    }
+    val newBuilderFunction = () => newBuilder
+    for (elem <- this.seq) {
```

Comment:
The existing code wasn't safe with parallel collections because of the absent `.seq` call and use of a mutable collection inside the loop.